pkg/agent/controller/policyrule/policyrule_controller.go

```diff
@@ -131,39 +129,25 @@ func (r *PolicyRuleReconciler) deletePolicyRule(e event.DeleteEvent, q workqueue
 }
 
 func (r *PolicyRuleReconciler) processPolicyRuleDelete(ruleName string) {
-	r.flowKeyReferenceMapLock.Lock()
-	defer r.flowKeyReferenceMapLock.Unlock()
-
 	var flowKey = flowKeyFromRuleName(ruleName)
-	if r.flowKeyReferenceMap[flowKey] == nil {
-		// already deleted
+
+	if _, exist := r.flowKeyReferenceCache.Get(flowKey); !exist {
 		return
 	}
 
-	r.flowKeyReferenceMap[flowKey].Delete(ruleName)
-
-	if r.flowKeyReferenceMap[flowKey].Len() == 0 {
-		delete(r.flowKeyReferenceMap, flowKey)
-
-		klog.Infof("remove rule %s from datapath", flowKey)
-		r.deletePolicyRuleFromDatapath(flowKey)
-	}
+	r.flowKeyReferenceCache.Delete(flowKey)
+	klog.Infof("remove rule %s from datapath", flowKey)
+	r.deletePolicyRuleFromDatapath(flowKey)
```

Comment:
flowKeyReferenceMap maps flowKey and PolicyRuleNames.
Only all PolicyRules with the flowKey be deleted, the flow should be deleted.